File: gridmap/scripts/VisibleGrid.py
#!/usr/bin/python
#-*- encoding: utf8 -*-

#目前该节点效率较低，只能跑到2Hz
from FakeDetector import getCellList
from scipy.spatial.transform import Rotation as R
import rospy
import numpy as np
import math
import time
from gazebo_msgs.msg import LinkStates
from sensor_msgs.msg import CameraInfo
from nav_msgs.msg import OccupancyGrid

'''
#获得以原点为圆心、一定半径、一定幅角范围内（保证圆心角不超过180度）圆内的所有整点
#NOTE: ang_range中可以包含超过两个元素，但是所有元素必须能被包含在一个圆心角不超过180度的扇形内
def getPointInCircle(rad, ang_range):
    ang_max = max(ang_range)
    ang_min = min(ang_range)
    is_m1_in_range = None # -1是否被包含在幅角范围内
    if ang_max - ang_min <= math.pi:
        is_m1_in_range = False
    else:
        is_m1_in_range = True
        ang_range_new = [(i+2*math.pi if i<0 else i) for i in ang_range]
        ang_min = max(ang_range_new)-2*math.pi
        ang_max = min(ang_range_new)
    pt_list = []
    if rad < 0:
        return pt_list
    for x in range(-int(rad), int(rad)+1):
        y_max = math.sqrt(rad*rad - x*x)
        for y in range(-int(y_max), int(y_max)+1):
            ang = np.angle(x+y*1j) #算幅角很慢
            if (is_m1_in_range and (ang<ang_min or ang>ang_max)) or \
                    (not is_m1_in_range and ang>ang_min and ang<ang_max):
                pt_list.append([x, y])
            pt_list.append([x, y])
    return pt_list
'''


class VisibleGridNode:

    def __init__(self):
        rospy.init_node('visible_grid_node', anonymous=True)
        
        #地图信息和相机、目标的位姿
        self.grid_msg_ = None
        self.res_ = None
        self.origin_ = None
        self.map_width_ = None
        self.map_height_ = None
        self.t_wc_ = np.zeros([3], dtype = 'float')
        self.R_wc_ = R.from_quat([0, 0, 0, 1])
        self.is_grid_init_ = False
        self.is_camera_init_ = False

        #相机、目标model的名字
        self.camera_name_ = 'iris_0::base_link'

        #相机参数
        self.fx_ = 0.0
        self.fy_ = 0.0
        self.cx_ = 0.0
        self.cy_ = 0.0
        self.h_ = 0
        self.w_ = 0
        self.is_intrinsics_init_ = False

        self.max_dist_ = 30.0  #最远距离
        self.visible_grid_ = None

        self.gridPub_ = rospy.Publisher('/available_grid', OccupancyGrid, queue_size=100)

        self.gridSub_ = rospy.Subscriber('/occupancy_grid', OccupancyGrid, self.occupancygridCallback)
        self.linkposeSub_ = rospy.Subscriber('/gazebo/link_states', LinkStates, self.linkposeCallback)
        self.intrinsicsSub_ = rospy.Subscriber('/iris_0/usb_cam/camera_info', CameraInfo, self.intrinsicsCallback)

        self.publishloop_timer_ = rospy.Timer(rospy.Duration(10.0), self.publishloopCallback)

        try:
            rospy.spin()
        except KeyboardInterrupt:
            pass


    def gridUpdate(self):
        t1 = time.clock()
        if not self.is_camera_init_ or not self.is_intrinsics_init_ or not self.is_grid_init_:
            return
        pix_list = [np.array([1.0, (self.cx_ - 0.0) / self.fx_, (self.cy_ - 0.0) / self.fy_]), \
                    np.array([1.0, (self.cx_ - 0.0) / self.fx_, (self.cy_ - self.h_) / self.fy_]), \
                    np.array([1.0, (self.cx_ - self.w_) / self.fx_, (self.cy_ - 0.0) / self.fy_]), \
                    np.array([1.0, (self.cx_ - self.w_) / self.fx_, (self.cy_ - self.h_) / self.fy_])]   #相机图像四角的归一化坐标
        yaw_list = []
        for p_c in pix_list:
            p_w = np.matmul(self.R_wc_.as_dcm(), p_c)
            yaw_list.append(np.angle(p_w[0]+p_w[1]*1j))
        #获取扫描角度范围
        yaw_max = max(yaw_list)
        yaw_min = min(yaw_list)
        if yaw_max - yaw_min <=math.pi:
            pass
        else:
            yaw_list_new = [(yaw+2*math.pi if yaw<0 else yaw) for yaw in yaw_list]
            yaw_min = min(yaw_list_new)
            yaw_max = max(yaw_list_new)
        #开始扫描
        t2 = time.clock()
        self.visible_grid_ = np.zeros([self.map_width_, self.map_height_], dtype = np.uint8)
        yaw = yaw_min
        yaw_step = self.res_ / self.max_dist_
        R_wc_mat = self.R_wc_.inv().as_dcm()
        t_wc_vec = self.t_wc_.copy()
        while yaw <= yaw_max:
            endpoint = self.t_wc_[:2] + self.max_dist_ * np.array([math.cos(yaw), math.sin(yaw)])
            cell_list = getCellList(self.t_wc_[:2], endpoint, self.res_, self.origin_)
            for cell in cell_list:
                if cell[0] < 0 or cell[0] >= self.map_width_ or cell[1] < 0 or cell[1] >= self.map_height_:
                    break
                if self.grid_msg_.data[cell[1]*self.map_width_+cell[0]] > 0:
                    break
                #判断该点是否在相机视野内
                '''
                t_wt = self.origin_ + (np.array(cell, dtype = 'float') + np.array([0.5, 0.5])) * self.res_
                t_wt = np.append(t_wt, 0)
                t_ct = np.matmul(self.R_wc_.inv().as_dcm(), (t_wt - self.t_wc_))
                X = -t_ct[1]
                Y = -t_ct[2]
                Z = t_ct[0]
                u = self.fx_*(X/Z) + self.cx_
                v = self.fy_*(Y/Z) + self.cy_
                '''
                
                t_wt_x = self.origin_[0] + (float(cell[0]) + 0.5) * self.res_
                t_wt_y = self.origin_[1] + (float(cell[1]) + 0.5) * self.res_
                t_ct = np.matmul(R_wc_mat, (np.array([t_wt_x, t_wt_y, 0.0]) - t_wc_vec))
                u = self.fx_*(-t_ct[1]/t_ct[0]) + self.cx_
                v = self.fy_*(-t_ct[2]/t_ct[0]) + self.cy_
                
                if t_ct[0] < 0 or u < 0 or u > float(self.w_) or v < 0 or v > float(self.h_):
                    continue
                
                self.visible_grid_[cell[0]][cell[1]] = 100
            yaw += yaw_step
        t3 = time.clock()
        print((t2-t1), (t3-t2))


    def occupancygridCallback(self, msg):
        if not self.is_grid_init_:
            self.grid_msg_ = msg
            self.res_ = msg.info.resolution
            self.origin_ = np.array([msg.info.origin.position.x, msg.info.origin.position.y])
            self.map_width_ = msg.info.width
            self.map_height_ = msg.info.height
            self.is_grid_init_ = True


    def linkposeCallback(self, msg):
        if self.camera_name_ in msg.name:
            i = msg.name.index(self.camera_name_)
            pose = msg.pose[i]
            self.t_wc_ = np.array([pose.position.x, pose.position.y, pose.position.z])
            self.R_wc_ = R.from_quat([pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w])
            self.is_camera_init_ = True


    def intrinsicsCallback(self, msg):
        self.fx_ = msg.K[0]
        self.fy_ = msg.K[4]
        self.cx_ = msg.K[2]
        self.cy_ = msg.K[5]
        self.h_ = msg.height
        self.w_ = msg.width
        self.is_intrinsics_init_ = True


    def publishloopCallback(self, event):
        self.gridUpdate()
        if self.visible_grid_ is not None:
            grid_msg = OccupancyGrid()
            grid_msg.header.stamp = rospy.Time.now()
            grid_msg.info.map_load_time = rospy.Time.now()
            grid_msg.info.resolution = self.res_
            grid_msg.info.width = self.map_width_
            grid_msg.info.height = self.map_height_
            grid_msg.info.origin.position.x = self.origin_[0]
            grid_msg.info.origin.position.y = self.origin_[1]
            grid_msg.info.origin.position.z = 0
            grid_msg.info.origin.orientation.x = 0
            grid_msg.info.origin.orientation.y = 0
            grid_msg.info.origin.orientation.z = 0
            grid_msg.info.origin.orientation.w = 1
            map_result_list = self.visible_grid_.T.flatten().tolist()
            for ele in map_result_list:
                ele = int(ele)
            grid_msg.data = map_result_list
            self.gridPub_.publish(grid_msg)


if __name__ == '__main__':
    vg = VisibleGridNode()

File: gridmap/src/VisibleGrid.cpp
#include "VisibleGrid.h"
#include <iostream>
#include <sys/time.h>
//#include <xmlrpcpp/XmlRpcValue.h>
using namespace std;

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

vector<pairii> getCellLists(Eigen::Vector2d pt1, Eigen::Vector2d pt2, double res, Eigen::Vector2d origin)
{
    Eigen::Vector2d pt1_ref = (pt1 - origin) / res;
    Eigen::Vector2d pt2_ref = (pt2 - origin) / res;
    double x1 = pt1_ref[0], y1 = pt1_ref[1];
    double x2 = pt2_ref[0], y2 = pt2_ref[1];
    vector<pairii> cell_list;
    if (fabs(x1-x2) < 1e-6 or floor(x1) == floor(x2))
    {
        if (y1 <= y2)
        {
            for (int y = int(floor(y1)); y <= int(floor(y2)); y++)
            {
                cell_list.push_back(pairii(int(floor(x1)), y));
            }
        }
        else
        {
            for (int y = int(floor(y1)); y >= int(floor(y2)); y--)
            {
                cell_list.push_back(pairii(int(floor(x1)), y));
            }
        }
    }
    else if (fabs(y1-y2) < 1e-6 or floor(y1) == floor(y2))
    {
        if (x1 <= x2)
        {
            for (int x = int(floor(x1)); x <= int(floor(x2)); x++)
            {
                cell_list.push_back(pairii(x, int(floor(y1))));
            }
        }
        else
        {
            for (int x = int(floor(x1)); x >= int(floor(x2)); x--)
            {
                cell_list.push_back(pairii(x, int(floor(y1))));
            }
        }
    }
    else
    {
        // 思路：求解线段与格线的所有交点，进而得到线段经过的所有格子
        // y=kx+b, x=ly+c
        bool inv;
        if (floor(x1) > floor(x2))
        {
            int temp;
            inv = true;
            temp = x1; x1 = x2; x2 = temp;
            temp = y1; y1 = y2; y2 = temp;
        }
        else
        {
            inv = false;
        }
        // 此后均按照x1<x2情况处理，cell_list从左往右
        vector<pairdd> intersect_list_x, intersect_list_y;
        double k = (y2-y1) / (x2-x1);
        double b = (x2*y1-x1*y2) / (x2-x1);
        for (int x = int(floor(x1))+1; x <= int(floor(x2)); x++)
        {
            intersect_list_x.push_back(pairdd(double(x), k*double(x)+b));
        }
        if (k > 0)
        {
            for (int y = int(floor(y1))+1; y <= int(floor(y2)); y++)
            {
                intersect_list_y.push_back(pairdd((double(y)-b)/k, double(y)));
            }
        }
        else
        {
            for (int y = int(floor(y1)); y >= int(floor(y2))+1; y--)
            {
                intersect_list_y.push_back(pairdd((double(y)-b)/k, double(y)));
            }
        }
        int ptrx = 0, ptry = 0;
        bool isX;
        while (ptrx < intersect_list_x.size() || ptry < intersect_list_y.size())  // 改用一次归并排序
        {
            if (ptry >= intersect_list_y.size())
            {
                isX = true;
            }
            else if (ptrx >= intersect_list_x.size())
            {
                isX = false;
            }
            else
            {
                isX = intersect_list_x[ptrx].first < intersect_list_y[ptry].first;
            }
            if (isX)
            {
                cell_list.push_back(pairii(int(intersect_list_x[ptrx].first)-1, int(floor(intersect_list_x[ptrx].second))));
                ptrx++;
            }
            else
            {
                if (k > 0)
                {
                    cell_list.push_back(pairii(int(floor(intersect_list_y[ptry].first)), int(intersect_list_y[ptry].second)-1));
                }
                else
                {
                    cell_list.push_back(pairii(int(floor(intersect_list_y[ptry].first)), int(intersect_list_y[ptry].second)));
                }
                ptry++;
            }
        }
        cell_list.push_back(pairii(int(floor(x2)), int(floor(y2))));
        if (inv)
        {
            reverse(cell_list.begin(), cell_list.end());
        }
    }
    return cell_list;
}

VisibleGridNode::VisibleGridNode()
{
    this->is_grid_init_ = false;
    this->is_camera_init_ = false;

    this->camera_name_ = "iris_0::base_link";
    
    this->is_intrinsics_init_ = false;

    this->max_dist_ = 30.0;

    this->gridPub_ = n_.advertise<nav_msgs::OccupancyGrid>("/available_grid", 100);

    this->gridSub_ = n_.subscribe("/occupancy_grid", 1, &VisibleGridNode::occupancygridCallback, this, ros::TransportHints().tcpNoDelay());  
    this->linkposeSub_ = n_.subscribe("/gazebo/link_states", 1, &VisibleGridNode::linkposeCallback, this, ros::TransportHints().tcpNoDelay());  
    this->intrinsicsSub_ = n_.subscribe("iris_0/usb_cam/camera_info", 1, &VisibleGridNode::intrinsicsCallback, this, ros::TransportHints().tcpNoDelay());

    this->publishloop_timer_ = n_.createTimer(ros::Duration(0.5), &VisibleGridNode::publishloopCallback, this);

}

void VisibleGridNode::gridUpdate()
{    
    struct timeval t1, t2;
    gettimeofday(&t1, NULL);

    if (!this->is_camera_init_ || !this->is_intrinsics_init_ || !this->is_grid_init_)
    {
        return;
    }
    Eigen::Vector3d pix_list[4] = {Eigen::Vector3d(1.0, (this->cx_ - 0.0) / this->fx_, (this->cy_ - 0.0) / this->fy_), 
                                   Eigen::Vector3d(1.0, (this->cx_ - 0.0) / this->fx_, (this->cy_ - this->h_) / this->fy_), 
                                   Eigen::Vector3d(1.0, (this->cx_ - this->w_) / this->fx_, (this->cy_ - 0.0) / this->fy_), 
                                   Eigen::Vector3d(1.0, (this->cx_ - this->w_) / this->fx_, (this->cy_ - this->h_) / this->fy_)};
    double yaw_list[4];
    double yaw_max = -10, yaw_min = 10;
    for (int i = 0; i < 4; i++)
    {
        Eigen::Vector3d p_w = this->R_wc_ * pix_list[i];
        yaw_list[i] = atan2(p_w[1], p_w[0]);
        if (yaw_list[i] > yaw_max)
        {
            yaw_max = yaw_list[i];
        }
        if (yaw_list[i] < yaw_min)
        {
            yaw_min = yaw_list[i];
        }
    }
    if (yaw_max - yaw_min > M_PI)
    {
        yaw_max = -10;
        yaw_min = 10;
        for (int i = 0; i < 4; i++)
        {
            if (yaw_list[i] < 0)
            {
                yaw_list[i] += 2*M_PI;
            }
            if (yaw_list[i] > yaw_max)
            {
                yaw_max = yaw_list[i];
            }
            if (yaw_list[i] < yaw_min)
            {
                yaw_min = yaw_list[i];
            }
        }
    }
    //if (this->visible_grid_.size() == 0)
    if (true)
    {
        this->visible_grid_ = vector<signed char>(this->map_width_ * this->map_height_);
    }
    double yaw = yaw_min;
    double yaw_step = this->res_ / this->max_dist_;
    Eigen::Quaterniond R_wc_inv = this->R_wc_.conjugate();
    Eigen::Vector3d t_wc_vec = this->t_wc_;
    while (yaw <= yaw_max)
    {
        Eigen::Vector2d endpoint;
        Eigen::Vector2d t_wc_xy = Eigen::Vector2d(t_wc_vec[0], t_wc_vec[1]);
        endpoint = t_wc_xy + this->max_dist_ * Eigen::Vector2d(cos(yaw), sin(yaw));
        vector<pairii> cell_list = getCellLists(t_wc_xy, endpoint, this->res_, this->origin_);
        for (auto iter = cell_list.begin(); iter != cell_list.end(); iter++)
        {
            if (iter->first < 0 || iter->first >= this->map_width_ || iter->second < 0 || iter->second >= this->map_height_)
            {
                break;
            }
            if (this->grid_msg_.data[iter->second*this->map_width_ + iter->first] > 0)
            {
                break;
            }
            double t_wt_x = this->origin_[0] + (double(iter->first) + 0.5) * this->res_;
            double t_wt_y = this->origin_[1] + (double(iter->second) + 0.5) * this->res_;
            Eigen::Vector3d t_ct = R_wc_inv * (Eigen::Vector3d(t_wt_x, t_wt_y, 0.0) - t_wc_vec);
            double u = this->fx_ * (-t_ct[1]/t_ct[0]) + this->cx_;
            double v = this->fy_ * (-t_ct[2]/t_ct[0]) + this->cy_;

            if (t_ct[0] < 0 || u < 0 || u > double(this->w_) || v < 0 || v > double(this->h_))
            {
                continue;
            }
            this->visible_grid_[iter->second*this->map_width_ + iter->first] = 100;
        }
        yaw += yaw_step;
    }

    gettimeofday(&t2, NULL);
    double deltaT = (t2.tv_sec - t1.tv_sec) * 1000000 + t2.tv_usec - t1.tv_usec;
    std::cout << "time consumed "<< deltaT << " us" <<std::endl;
}

void VisibleGridNode::occupancygridCallback(const nav_msgs::OccupancyGrid::ConstPtr& msg)
{
    if (!this->is_grid_init_)
    {
        this->grid_msg_ = *msg;
        this->res_ = msg->info.resolution;
        this->origin_ = Eigen::Vector2d(msg->info.origin.position.x, msg->info.origin.position.y);
        this->map_width_ = msg->info.width;
        this->map_height_ = msg->info.height;
        this->is_grid_init_ = true;
    }
}

void VisibleGridNode::linkposeCallback(const gazebo_msgs::LinkStates::ConstPtr& msg)
{
    int i;
    for(i = 0; i < msg->name.size(); i++)
    {
        if(msg->name[i] == this->camera_name_)
        {
            break;
        }
    }
    if(i < msg->name.size())
    {
        geometry_msgs::Pose pose = msg->pose[i];
        this->t_wc_ << pose.position.x, pose.position.y, pose.position.z;
        this->R_wc_ = Eigen::Quaterniond(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
        this->is_camera_init_ = true;
    }
}

void VisibleGridNode::intrinsicsCallback(const sensor_msgs::CameraInfo::ConstPtr& msg)
{
    this->fx_ = msg->K[0];
    this->fy_ = msg->K[4];
    this->cx_ = msg->K[2];
    this->cy_ = msg->K[5];
    this->h_ = msg->height;
    this->w_ = msg->width;
    this->is_intrinsics_init_ = true;
}

void VisibleGridNode::publishloopCallback(const ros::TimerEvent& event)
{
    if (this->visible_grid_.size() != 0)
    {
        nav_msgs::OccupancyGrid grid;
        grid.header.stamp = ros::Time::now();
        grid.info.map_load_time = ros::Time::now();
        grid.info.resolution = this->res_;
        grid.info.width = this->map_width_;
        grid.info.height = this->map_height_;
        grid.info.origin.position.x = this->origin_[0];
        grid.info.origin.position.y= this->origin_[1];
        grid.info.origin.position.z = 0;
        grid.info.origin.orientation.x = 0;
        grid.info.origin.orientation.y = 0;
        grid.info.origin.orientation.z = 0;
        grid.info.origin.orientation.w = 1;
        grid.data = this->visible_grid_;
        this->gridPub_.publish(grid);
    }
}

int main(int argc, char **argv)
{   
/*
    Eigen::Vector2d start(600, 500), end(800, 498.1), origin(0, 0);
    vector<pairii> cell_list = getCellLists(start, end, 1, origin);
    for (auto iter = cell_list.begin(); iter != cell_list.end(); iter++)
        cout << "(" << iter->first << "  " << iter->second << ")\n";
*/
    ros::init(argc, argv, "visible_grid_node");

    VisibleGridNode f;

    //ros::Rate loop_rate(100);
    while(ros::ok())
    {
        f.gridUpdate();
        ros::spinOnce();
        //loop_rate.sleep();
    }

    return 0;
}
